x-pack/plugins/security_solution/server/endpoint/routes/resolver/queries/stats.ts

```diff
@@ -112,34 +130,106 @@ export class StatsQuery extends ResolverQuery<StatsResult> {
         alerts: {
           filter: { term: { 'event.kind': 'alert' } },
           aggs: {
-            ids: { terms: { field: 'process.entity_id' } },
+            ids: { terms: { field: 'process.entity_id', size: entityIDs.length } },
           },
         },
         events: {
           filter: { term: { 'event.kind': 'event' } },
           aggs: {
-            ids: { terms: { field: 'process.entity_id' } },
+            ids: {
+              // The entityIDs array will be made up of alert and event entity_ids, so we're guaranteed that there
+              // won't be anymore unique process.entity_ids than the size of the array passed in
+              terms: { field: 'process.entity_id', size: entityIDs.length },
+              aggs: {
+                categories: {
+                  // Currently ECS defines a small number of valid categories (under 10 right now), as ECS grows it's possible that the
+                  // valid categories could exceed this hardcoded limit. If that happens we might want to revisit this
+                  // and transition it to a composite aggregation so that we can paginate through all the possible response
+                  terms: { field: 'event.category', size: 1000 },
+                },
+              },
+            },
           },
         },
       },
     };
   }
 
-  public formatResponse(response: SearchResponse<ResolverEvent>): StatsResult {
-    const alerts = response.aggregations.alerts.ids.buckets.reduce(
-      (cummulative: Record<string, number>, bucket: AggBucket) => ({
-        ...cummulative,
-        [bucket.key]: bucket.doc_count,
-      }),
-      {}
-    );
-    const events = response.aggregations.events.ids.buckets.reduce(
+  private static getEventStats(catAgg: CategoriesAgg): EventStats {
+    const total = catAgg.doc_count;
+    if (!catAgg.categories?.buckets) {
+      return {
+        total,
+        categories: {},
+      };
+    }
+
+    const categories: Record<string, number> = catAgg.categories.buckets.reduce(
       (cummulative: Record<string, number>, bucket: AggBucket) => ({
         ...cummulative,
         [bucket.key]: bucket.doc_count,
       }),
       {}
     );
+    return {
+      total,
+      categories,
+    };
+  }
+
+  public formatResponse(response: SearchResponse<ResolverEvent>): StatsResult {
+    let alerts: Record<string, number> = {};
+
+    if (response.aggregations?.alerts?.ids?.buckets) {
+      alerts = response.aggregations.alerts.ids.buckets.reduce(
+        (cummulative: Record<string, number>, bucket: AggBucket) => ({
+          ...cummulative,
+          [bucket.key]: bucket.doc_count,
```

Comment:
what is the reason for changing the shape of this?

Reply:
So this particular `.reduce` shouldn't have actually changed. I just wrapped it in an `if` check because `response.aggregations` is an `any` type and I'm not sure if the `alerts` aggregation will always be defined. The `reduce` call should be exactly the same though.